### PR TITLE
Pin update March 2024

### PR DESCRIPTION
**Pull request by lsy323**
Update xla pin to HEAD

Summary:
- Update bazel to 6.5.0
- Rename `PJRT_Structure_Base` to `PJRT_Extension_Base` to accommodate change in XLA. 
- Generate custom_call to `mhlo.uniform_dequantize/quantize` to accommodate the change in HLO->MHLO converter patch.
- Reduce the BS of several GPU tests to unblock the pin update from the OOM issue in GPU tests. (Details in comment below).

### CODE REVIEW

**Comment by lsy323** [comment]
Hit the following error
```
 File "/home/lsiyuan/.cache/bazel/_bazel_lsiyuan/9d8c0c9d904275861907f86bf4a21dbc/external/llvm-project/mlir/BUILD.bazel", line 40, column 7, in <toplevel>
                } | if_cuda_available(
Error: unsupported binary operation: dict | select
```

Need to upgrade bazel version to above 6.0.0

**Comment by lsy323** [comment]
Testing performance with
```

```

After pin update
```
| Training Device=xla:0/1 Epoch=1 Step=2280 Loss=0.00135 Rate=425.92 GlobalRate=370.64 Time=23:07:35
| Training Device=xla:0/2 Epoch=1 Step=2280 Loss=0.00135 Rate=425.92 GlobalRate=371.41 Time=23:07:35
| Training Device=xla:0/3 Epoch=1 Step=2280 Loss=0.00135 Rate=425.92 GlobalRate=371.07 Time=23:07:35
| Training Device=xla:0/1 Epoch=1 Step=2300 Loss=0.00135 Rate=425.11 GlobalRate=371.04 Time=23:07:41
| Training Device=xla:0/2 Epoch=1 Step=2300 Loss=0.00135 Rate=425.12 GlobalRate=371.81 Time=23:07:41
| Training Device=xla:0/0 Epoch=1 Step=2300 Loss=0.00135 Rate=425.12 GlobalRate=371.80 Time=23:07:41
| Training Device=xla:0/3 Epoch=1 Step=2300 Loss=0.00135 Rate=425.12 GlobalRate=371.48 Time=23:07:41
```

**Comment by lsy323** [comment]
Test failed with PT2E test, move xla pin again after https://github.com/pytorch/xla/blob/master/openxla_patches/quant_dequant_converter.diff is upstreamed

**Comment by lsy323** [comment]
The following GPU tests hit OOM in CI after pin update

```
PJRT_DEVICE=CUDA torchrun --nnodes=1 --node_rank=0 --nproc_per_node=2 test/test_train_mp_imagenet.py --fake_data --pjrt_distributed --batch_size=16 --num_epochs=1 --num_steps=25 --model=resnet18

PJRT_DEVICE=CUDA python test/test_train_mp_imagenet_fsdp.py --fake_data --auto_wrap_policy type_based --use_small_fake_sample --num_epochs=1

PJRT_DEVICE=CUDA python test/test_train_mp_imagenet_fsdp.py --fake_data --use_nested_fsdp --use_small_fake_sample --num_epochs=1
```

File: WORKSPACE [hunk 1]
```diff
@@ -67,7 +67,7 @@ http_archive(
 # b) by commenting out the http_archive above and uncommenting the following:
 # local_repository(
 #    name = "xla",
-#    path = "/path/to/openxla",
+#    path = "/home/lsiyuan/work/xla",
```
Comment (frgossen) [inline WORKSPACE:70]: Is this intended?

Comment (lsy323) [reply WORKSPACE:70]: Not intended, for local testing purpose. I'll recover it.

**Comment by lsy323** [comment]
cc @will-cromar for some change on the PJRT interface.

**Comment by lsy323** [comment]
Thanks @sdasgup3 for pointing out we need to generate custom call to `mhlo.uniform_de/quantize`. to accommodate the incoming `stablehlo.uniform_quantize/dequantize` in HLO->MHLO converter.

**Review by JackCaoG — APPROVED** [review]

**Review by yeounoh — APPROVED** [review]
LGTM

**Pull request merged**